Add custom events (#683)
OffchainLabs/arbitrum-docs: arbitrum-docs/launch-orbit-chain/how-tos/customize-precompile.mdx

hi
```

## Option 3: Define a new event

> Here we will still use the precompile `Arbsys` as an example and add a simple `Hi` event to be emitted in `SayHi` method which we added in `ArbSys.sol` contract at Option 1, first go to the [precompiles implementation][precompile_impl_dir_link] directory, and find `ArbSys.go`, edit the `ArbSys` struct:

Reviewer: ```suggestion
Here we will use the same precompile `Arbsys` as an example and add a simple `Hi` event to be emitted in the `SayHi` method which we added in `ArbSys.sol` contract in Option 1. First, navigate to the [precompiles implementation][precompile_impl_dir_link] directory and find the `ArbSys.go`. Next, edit the `ArbSys` struct:
```

```go
// ArbSys provides system-level functionality for interacting with L1 and understanding the call stack.
type ArbSys struct {
	Address                 addr // 0x64
	L2ToL1Tx                func(ctx, mech, addr, addr, huge, huge, huge, huge, huge, huge, []byte) error
	L2ToL1TxGasCost         func(addr, addr, huge, huge, huge, huge, huge, huge, []byte) (uint64, error)
	SendMerkleUpdate        func(ctx, mech, huge, bytes32, huge) error
	SendMerkleUpdateGasCost func(huge, bytes32, huge) (uint64, error)
	InvalidBlockNumberError func(huge, huge) error

	// deprecated event
	L2ToL1Transaction        func(ctx, mech, addr, addr, huge, huge, huge, huge, huge, huge, huge, []byte) error
	L2ToL1TransactionGasCost func(addr, addr, huge, huge, huge, huge, huge, huge, huge, []byte) (uint64, error)

	// Add your customize event here:
	Hi						func(ctx, mech, addr) error

> 	// This is needed which will tell you how much gas it will cost, the param is the same as your event but without the first two (ctx, mech), the return param is always (uint64, error)

Reviewer: ```suggestion
	// This is needed and will tell you how much gas it will cost, the param is the same as your event but without the first two (ctx, mech), the return param is always (uint64, error)
```

	HiGasCost				func(addr) (uint64, error)
}
```

> Then add the event to `SayHi` method:

Reviewer: ```suggestion
Then add the event to the `SayHi` method:
```

```go
func (con *ArbSys) SayHi(c ctx, evm mech) (string, error) {
	err := con.Hi(c, evm, c.caller)
	return "hi", err
}
```

> Now navigate to the [precompiles interface][precompile_interface_dir_link] directory, open `Arbsys.sol`, and add the required interface. Ensure that the Event name on the interface matches the name of the function you introduced in `ArbSys` struct in the previous step:

Reviewer: ```suggestion
Now navigate to the [precompiles interface][precompile_interface_dir_link] directory, open `Arbsys.sol`, and add the required interface. Ensure that the event name on the interface matches the name of the function you introduced in `ArbSys` struct in the previous step:
```

```solidity
event Hi(address caller);
```

> If you want to make the param in event as `index`, just add index to solidity interface as this:

Reviewer: ```suggestion
To specify the event parameter as `index`, just include it in the Solidity interface like this:
```

```solidity

> event Hi(address indexed caller);

Author: indexed is a keyword in solidity, so we shouldn't change to this

```

> Now as emiting events need to cost gas, so we should remove `view` function behavior from the method:

Reviewer: ```suggestion
You need to remove the 'view' function behavior from the method as emitting events consumes gas.
```

```solidity
function sayHi() external returns(string memory);
```

> Next, build Nitro by following the instructions in [How to build Nitro locally](/node-running/how-tos/build-nitro-locally). Note that if you've already built the Docker image, you still need run the last step to rebuild.

Reviewer: ```suggestion
Next, build Nitro using the the instructions in [How to build Nitro locally](/node-running/how-tos/build-nitro-locally). Note that if you've already built the Docker image, you still need run the last step to rebuild.
```

Run Nitro with the following command:

```shell
docker run --rm -it  -v /some/local/dir/arbitrum:/home/user/.arbitrum -p 0.0.0.0:8547:8547 -p 0.0.0.0:8548:8548 @latestNitroNodeImage@ --parent-chain.connection.url=<YourParentChainUrl> --chain.id=<YourOrbitChainId> --http.api=net,web3,eth,debug --http.corsdomain=* --http.addr=0.0.0.0 --http.vhosts=*
```

:::info

Note that the instructions provided in [How to run a full node](/node-running/how-tos/running-a-full-node) **will not** work with your Orbit node. See [Command-line options (Orbit)](/launch-orbit-chain/reference/command-line-options) for Orbit-specific CLI flags.

:::

### Send the transaction and get the transaction receipt

> Send transaction to `ArbSys`, note as the function is not view/pure function, we need to send transaction with gas cost:

Reviewer: ```suggestion
Send a transaction to `ArbSys`, note that since this isn't a view or pure function, you'll need to send a transaction with associated gas costs:
```

```
cast send 0x0000000000000000000000000000000000000064 "sayHi()(string)"
```

> Then we will have a transaction hash result, call `eth_getTransactionReceipt` with that transaction hash, you may get:

Reviewer: ```suggestion
Then you will have a transaction hash result. Calling`eth_getTransactionReceipt` with the transaction hash will give you the following:
```

```

> You should see `logs` field within the transaction receipt like this:

Reviewer 2: ```suggestion
Note the `logs` field within the transaction receipt:
```

```
"logs":[
         {
            "address":"0x0000000000000000000000000000000000000064",
            "topics":[
               "0xa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa",
               "0x000000000000000000000000{Your_address}"
            ],
            "data":"0x",
            "blockNumber":"0x40",
            "transactionHash":"{Your_txHash}",
            "transactionIndex":"0x1",
            "blockHash":"0x0b367d705002b3575db99354a0964c033f929f26f4442ed347e47ae43a8f28e4",
            "logIndex":"0x0",
            "removed":false
         }
      ]
```

> From the `logs` field, we can see the topics[0] is `0xa937..cfa`, which is the event signature of `Hi(address)`, you can check it using [4 byte directory](https://www.4byte.directory/event-signatures/?bytes_signature=0xa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa), and topics[1] is `Your_address` which is exactly what we defined above! 

Reviewer: ```suggestion
From the `logs` field, you could see the topics[0] is `0xa937..cfa`, which is the event signature of `Hi(address)`, you can check it using [4 byte directory](https://www.4byte.directory/event-signatures/?bytes_signature=0xa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa), and topics[1] is `Your_address` which is exactly what we defined above! 
```

> From the `logs` field, we can see the topics[0] is `0xa937..cfa`, which is the event signature of `Hi(address)`, you can check it using [4 byte directory](https://www.4byte.directory/event-signatures/?bytes_signature=0xa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa), and topics[1] is `Your_address` which is exactly what we defined above! 

Reviewer 2: ```suggestion
From the `logs` field, `topics[0]` is `0xa937..cfa`, which is the event signature of `Hi(address)`. You can verify this using [4 byte directory](https://www.4byte.directory/event-signatures/?bytes_signature=0xa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa) - `topics[1]` is `Your_address`, which is exactly what we defined above! 
```

	SendMerkleUpdateGasCost func(huge, bytes32, huge) (uint64, error)
	InvalidBlockNumberError func(huge, huge) error

	// deprecated event
	L2ToL1Transaction        func(ctx, mech, addr, addr, huge, huge, huge, huge, huge, huge, huge, []byte) error
	L2ToL1TransactionGasCost func(addr, addr, huge, huge, huge, huge, huge, huge, huge, []byte) (uint64, error)

	// Add your customize event here:
	Hi						func(ctx, mech, addr) error
	// This is needed and will tell you how much gas it will cost, the param is the same as your event but without the first two (ctx, mech), the return param is always (uint64, error)
	HiGasCost				func(addr) (uint64, error)
}
```

Then add the event to the `SayHi` method:

```go
func (con *ArbSys) SayHi(c ctx, evm mech) (string, error) {
	err := con.Hi(c, evm, c.caller)
	return "hi", err
}
```

Now navigate to the [precompiles interface][precompile_interface_dir_link] directory, open `Arbsys.sol`, and add the required interface. Ensure that the event name on the interface matches the name of the function you introduced in `ArbSys` struct in the previous step:

```solidity
event Hi(address caller);
```

> If you want to make the param in event as `index`, just add `index` to the Solidity interface:

Reviewer 3: ```suggestion
If you want to [index the parameter](https://docs.soliditylang.org/en/latest/contracts.html#events) of the event (if you want to filter by that parameter in the future, for example), just add `indexed` to the Solidity interface:
```

Adding some more context to it. Feel free to modify !

```solidity
event Hi(address indexed caller);
```

> When an event is emitted, state is changed and a gas cost is incurred, so we should remove the `view` function behavior:

Reviewer 3: ```suggestion
Our function now emits an event, which means that when calling it, the state will change and a gas cost will be incurred. So we have to remove the `view` function behavior:
```

Small suggestion to clarify what this means. Modify as you want, or disregard if you don't think it's necessary to add more details to it!
